code/modules/atmospherics/machinery/components/machines/crystallizer.dm

```diff
@@ -0,0 +1,332 @@
+		quality_loss = max(quality_loss - 5.5, 100)
+
+	if(recipe[META_RECIPE_REACTION_TYPE] == "endothermic")
+		internal.temperature = max(internal.temperature - (recipe[META_RECIPE_ENERGY_RELEASE] / internal.heat_capacity()), TCMB)
+	else if(recipe[META_RECIPE_REACTION_TYPE] == "exothermic")
+		internal.temperature = max(internal.temperature + (recipe[META_RECIPE_ENERGY_RELEASE] / internal.heat_capacity()), TCMB)
+
+///Calculate the total moles needed for the recipe
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/moles_calculations()
+	var/amounts = 0
+	var/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	for(var/gas_type in recipe[META_RECIPE_REQUIREMENTS])
+		amounts += recipe[META_RECIPE_REQUIREMENTS][gas_type]
+	total_recipe_moles = amounts
+
+///Removes the gases from the internal gasmix when the recipe is changed
+/obj/machinery/atmospherics/components/binary/crystallizer/proc/dump_gases()
+	var/datum/gas_mixture/remove = internal.remove(internal.total_moles())
+	airs[2].merge(remove)
+	internal.garbage_collect()
+
+/obj/machinery/atmospherics/components/binary/crystallizer/process_atmos()
+	if(!on || !is_operational || recipe_type == null)
+		return
+
+	if(!check_gas_requirements())
+		return
+
+	inject_gases()
+
+	if(!internal.total_moles())
+		update_parents()
+		return
+	internal_check()
+
+	var/list/list/recipe = GLOB.gas_recipe_meta[recipe_type]
+	if(gas_check == recipe[META_RECIPE_REQUIREMENTS].len)
+		if(check_temp_requirements())
+			heat_calculations()
+			progress_bar = min(progress_bar + (MIN_PROGRESS_AMOUNT * 5 / (round(log(10, total_recipe_moles * 0.1), 0.01))), 100)
+		else
+			quality_loss = min(quality_loss + 0.5, 100)
+			progress_bar = max(progress_bar - 1, 0)
+	if(progress_bar != 100)
+		update_parents()
+		return
+	progress_bar = 0
+
+	for(var/gas_type in recipe[META_RECIPE_REQUIREMENTS])
+		var/amount_consumed = recipe[META_RECIPE_REQUIREMENTS][gas_type] + quality_loss * 5
+		if(internal.gases[gas_type][MOLES] < amount_consumed)
+			quality_loss = min(quality_loss + 10, 100)
+		internal.remove_specific(gas_type, amount_consumed)
+
+	var/total_quality = clamp(50 - quality_loss, 0, 100)
+	var/quality_control
+	switch(total_quality)
+		if(100)
+			quality_control = "Masterwork"
+		if(95 to 99)
+			quality_control = "Supreme"
+		if(75 to 94)
+			quality_control = "Good"
+		if(65 to 74)
+			quality_control = "Decent"
+		if(55 to 64)
+			quality_control = "Average"
+		if(35 to 54)
+			quality_control = "Ok"
+		if(15 to 34)
+			quality_control = "Poor"
+		if(5 to 14)
+			quality_control = "Ugly"
+		if(1 to 4)
+			quality_control = "Cracked"
+		if(0)
+			quality_control = "Oh God why"
+
+	for(var/path in recipe[META_RECIPE_PRODUCTS])
+		var/amount_produced = recipe[META_RECIPE_PRODUCTS][path]
+		for(var/i in 1 to amount_produced)
+			var/obj/creation = new path(get_step(src, SOUTH))
+			creation.name = "[quality_control] [creation.name]"
+			if(istype(creation, /obj/machinery/power/supermatter_crystal/shard))
```

Comment:
Make this a recipe variable instead of strict check ?

Reply:
i'll change it